executor/calibrate_resource.go

```diff
@@ -99,7 +166,92 @@ func (e *calibrateResourceExec) Next(ctx context.Context, req *chunk.Chunk) erro
 
 	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
 	ctx = kv.WithInternalSourceType(ctx, kv.InternalTxnOthers)
+	if len(e.optionList) > 0 {
+		return e.dynamicCalibrate(ctx, req, exec)
+	}
+	return e.staticCalibrate(ctx, req, exec)
+}
+
+func (e *calibrateResourceExec) dynamicCalibrate(ctx context.Context, req *chunk.Chunk, exec sqlexec.RestrictedSQLExecutor) error {
+	startTs, endTs, err := e.parseCalibrateDuration()
+	if err != nil {
+		return err
+	}
+	startTime := startTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
+	endTime := endTs.In(e.ctx.GetSessionVars().Location()).Format("2006-01-02 15:04:05")
+
+	totalKVCPUQuota, err := getTiKVTotalCPUQuota(ctx, exec)
+	if err != nil {
+		return err
+	}
+	totalTiDBCPU, err := getTiDBTotalCPUQuota(ctx, exec)
+	if err != nil {
+		return err
+	}
+	rus, err := getRUPerSec(ctx, exec, startTime, endTime)
+	if err != nil {
+		return err
+	}
+	tikvCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tikv", startTime, endTime)
+	if err != nil {
+		return err
+	}
+	tidbCPUs, err := getComponentCPUUsagePerSec(ctx, exec, "tidb", startTime, endTime)
+	if err != nil {
+		return err
+	}
+	quotas := make([]float64, 0)
+	lowCount, tidbCPULowCount, tikvCPULowCount := 0, 0, 0
+	for idx, ru := range rus {
+		if idx >= len(tikvCPUs) || idx >= len(tidbCPUs) {
+			break
+		}
+		tikvQuota, tidbQuota := tikvCPUs[idx]/totalKVCPUQuota, tidbCPUs[idx]/totalTiDBCPU
+		// If one of the two cpu usage is greater than the `valuableUsageThreshold`, we can accept it.
+		// And if both are greater than the `lowUsageThreshold`, we can also accpet it.
+		if tikvQuota > valuableUsageThreshold || tidbQuota > valuableUsageThreshold {
+			quotas = append(quotas, ru/mathutil.Max(tikvQuota, tidbQuota))
+		} else if tikvQuota < lowUsageThreshold {
+			lowCount++
+			tikvCPULowCount++
+			if tidbQuota < lowUsageThreshold {
+				tidbCPULowCount++
+			}
+		} else if tidbQuota < lowUsageThreshold {
+			lowCount++
+			tidbCPULowCount++
+		} else {
+			quotas = append(quotas, ru/mathutil.Max(tikvQuota, tidbQuota))
+		}
+	}
+	if len(quotas) < 5 {
+		return errors.Errorf("there are too few metrics points available")
+	}
+	if float64(len(quotas))/float64(len(quotas)+lowCount) > percentOfPass {
+		sort.Slice(quotas, func(i, j int) bool {
+			return quotas[i] > quotas[j]
+		})
+		lowerBound := int(math.Round(float64(len(quotas)) * float64(discardRate)))
+		upperBound := len(quotas) - lowerBound
+		sum := 0.
+		for i := lowerBound; i < upperBound; i++ {
+			sum += quotas[i]
+		}
+		quota := sum / float64(upperBound-lowerBound)
+		req.AppendUint64(0, uint64(quota))
+	} else {
+		if tidbCPULowCount > 0 && tikvCPULowCount > 0 {
+			return errors.Errorf("The CPU utilizations of TiDB and TiKV are less than one tenth in some of the time")
```

Comment:
Suggest msg
"The workload in selected time window is too low, with which TiDB is unable to reach a capacity estimation. Please select another time window with higher workload, or calibrate resource by hardware instead. "